### ISSUE TITLE
Wrong type-mismatch on `match` with `break` (II)

### ISSUE THREAD

**Issue by kenmasumitsu** [rust-analyzer/rust-analyzer#11837]
The codes below, 

```
enum MyErr {
    Err1,
    Err2,
}


fn example_ng() {
    let value: Result<i32, MyErr> = Ok(3);

    loop {
        let ret = match value {
            Ok(value) => value,
            Err(ref err) => {
                // expected i32, found ()rust-analyzer[type-mismatch](https://rust-analyzer.github.io/manual.html#type-mismatch)
                match err {
                    MyErr::Err1 => break,
                    MyErr::Err2 => continue,
                };
            }
        };
        println!("ret: {}", ret);
    }
}
```

This shows the error diagnostic `expected i32, found ()rust-analyzer[type-mismatch](https://rust-analyzer.github.io/manual.html#type-mismatch)`.

However, if enclosing with brace, the error diagnostic is not shown.

```
fn example_ok() {
    let value: Result<i32, MyErr> = Ok(3);

    loop {
        let ret = match value {
            Ok(value) => value,
            Err(ref err) => {
                match err {
                    MyErr::Err1 => {
                        break;
                    }
                    MyErr::Err2 => {
                        continue;
                    }
                };
            }
        };
        println!("ret: {}", ret);
    }
}
```

**rust-analyzer version**: (eg. output of "Rust Analyzer: Show RA Version" command)

rust-analyzer version: bc08b8eff 2022-03-28 stable

**rustc version**: (eg. output of `rustc -V`)

rustc 1.59.0 (9d1b2106e 2022-02-23)

**relevant settings**: (eg. client settings, or environment variables like `CARGO`, `RUSTUP_HOME` or `CARGO_HOME`)

**Comment by lnicola** [comment]
Possibly related to #11814.

**Comment by lnicola** [comment]
Fixed in #11844
